jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -115,28 +116,91 @@ protected void doStop() throws Exception
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Stopping {}", this);
-        CloseEndPoints close_endps = new CloseEndPoints();
-        submit(close_endps);
-        close_endps.await(getStopTimeout());
-        CloseSelector close_selector = new CloseSelector();
-        submit(close_selector);
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
+
+        // Grab a copy of the current selector, so we can force close endpoints later
+        Selector selector;
+        try (Locker.Lock lock = _locker.lock())
+        {
+            selector = _selector;
+        }
+
+        boolean graceful = getStopTimeout()>0;
+        boolean closed = selector==null;
+        if (graceful && !closed)
+        {
+            // Schedule a graceful close
+            GracefulCloseAction close = new GracefulCloseAction();
+            submit(close);
+            closed = close.await(getStopTimeout());
+        }
+        
+        try
+        {
+            super.doStop();
+        }
+        finally
+        {
+            // Really stop the selector
+            try (Locker.Lock lock = _locker.lock())
+            {
+                if (selector!=null && selector.isOpen())
+                    selector.wakeup();
+                _selector = null;
+            }
+            Thread.yield();
+
+            // Close any left over endpoints
+            if (!closed && selector.isOpen())
+            {
+                // close all the endpoints directly
+                selector.keys().stream()
+                .filter(SelectionKey::isValid)
+                .map(SelectionKey::attachment)
+                .filter(EndPoint.class::isInstance)
+                .map(EndPoint.class::cast)
+                .filter(EndPoint::isOpen)
+                .forEach(EndPoint::close);
+                selector.close();
+            }
```

Comment:
This whole `finally` block is racy with `GracefulCloseAction`, so if `GracefulCloseAction` takes more than `stopTimeout`, it will still run when the `finally` block runs.
Calls such as `selector.keys()` throw `ClosedSelectorException` (a runtime exception) if the selector is closed, so we will throw an exception while instead the closing has been done successfully already.
Also, the logic is different from `GracefulCloseAction` because there the close was forwarded to the `Connection`.
And it's racy with both `Selector.isOpen()` and `EndPoint.isOpen()`; the latter may be eliminated, perhaps, as `EndPoint.close()` should be idempotent.

Basically the `finally` block is doing the same of `GracefulCloseAction`, so perhaps we just call `run()` on it from the `finally` block, with an atomic guard in `run()` to run it just once: if it did not run it will now and skip it later, if it's already running it won't run again and it will eventually finish. WDYT ?

Reply:
The close block is deliberately different to the GracefulCloseAction.  Graceful close can call Connection.close, which may take time and send close handshakes etc. which may block.  None Graceful close is deliberately straight to the endpoint to close the underlying TCP connection regardless of protocol state.

However, I agree with the race concerns and we should protect with something atomic.   This is partially done by calling doStop before the finally block, but is perhaps not rigorous enough.

Reply:
Added an atomic reference to the selector to prevent the race